clippy_lints/src/too_owned.rs

```diff
@@ -0,0 +1,92 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_opt;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects needlessly owned `Cow` types.
+    ///
+    /// ### Why is this bad?
+    /// The borrowed types are usually more flexible (e.g. `&str` vs. `&String`).
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let wrogn: std::borrow::Cow<'_, String>;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let right: std::borrow::Cow<'_, str>;
+    /// ```
+    #[clippy::version = "1.85.0"]
+    pub OWNED_COW,
+    style,
+    "needlessly owned Cow type"
+}
+
+declare_lint_pass!(TooOwned => [OWNED_COW]);
+
+impl LateLintPass<'_> for TooOwned {
+    fn check_ty(&mut self, cx: &LateContext<'_>, ty: &hir::Ty<'_>) {
+        if let hir::TyKind::Path(qpath) = ty.kind
+            && let Some(cow_def_id) = cx.qpath_res(&qpath, ty.hir_id).opt_def_id()
+            && cx.tcx.is_diagnostic_item(sym::Cow, cow_def_id)
+            && let hir::QPath::Resolved(_, path) = qpath
+            && let [.., last_seg] = path.segments
+            && let Some(args) = last_seg.args
+            && let [_lt, carg] = args.args
+            && let hir::GenericArg::Type(cty) = carg
+            && let Some((span, repl)) = replacement(cx, cty)
```

Comment:
Might be worth adding a span checks like `!is_external_macro(...)` to avoid cases where the user can't do anything?

Reply:
The `types` checks already appear to do that.